client/src/pages/sw.ts

```diff
@@ -21,6 +25,81 @@ function* deleteAll(keys: Iterable<string>) {
             yield caches.delete(key);
 }
 
+async function requestBackgroundSync(ty: string) {
+    // We are offline, set a tag to insert snapshot.
+    await self.registration.sync.register(ty);
+    const tag = await self.registration.sync.getTags();
+    assert(tag.includes(ty));
+}
+
+async function pushEntriesToServer() {
+    // Load up all entries in localForage and create an array of promises that we will push to the server.
+    const promises = (await localForage.keys()).map(async key => {
+        const defer: DeferredFetch | null = await localForage.getItem(key);
+        assert(defer !== null);
+        return fetch(defer.url, {
+            credentials: defer.credentials,
+            method: defer.method,
+            body: JSON.stringify(defer.body),
+            headers: defer.headers,
+        });
+    });
+
+    // Clear storage and reset deferredSnaps store if all entries pushed.
+    await localForage.clear();
+
+    // Toss all to the server
+    const client = await self.clients.matchAll();
+    client.map(client=> client.postMessage('sync'));
+    return Promise.allSettled(promises);
+
+    // TODO: Error handler when atleast one fails.
+    // TODO: UI facing indicator of entries being pushed.
+}
+
+async function handleDocumentPost(req: Request): Promise<Response> {
+    try {
+        const res = await fetch(req.clone());
+        return res;
+    } catch (error) {
+        // Threw an error, you are probably offline.
+        assert(error instanceof TypeError);
+
+        const url = new URL(req.url);
+        let key = '';
+
+        // Extract document information and status from json body.
+        if (url.pathname.startsWith('/api/snapshot')) {
+            // Insert snapshot schema uses doc as barcode id.
+            const { doc } = DeferredSnapshotSchema.parse(await req.clone().json());
+            key = doc;
+        } else {
+            // Registration uses id to hold barcode id. It dosen't have a status either.
+            const { id } = DeferredRegistrationSchema.parse(await req.clone().json());
+            key = id;
+        }

```

Comment:
Let's use Rusty 🦀 code here.
```suggestion
        const { pathname } = new URL(req.url);
        const json = await req.clone().json();
        const key = pathname.startsWith('/api/snapshot')
            ? DeferredSnapshotSchema.parse(json).doc
            : DeferredRegistrationSchema.parse(json).id;
```